### PR TITLE
Fix overcounting recovered bytes after network disconnect 

### LINKED ISSUE #90441: Retrying a snapshot-based recovery may over-count `recovered_bytes` in stats
We saw a case where the recovery stats reported `recovered_bytes` greater than `length_in_bytes` for one file:

```
GET /_recovery?detailed

200 OK
...
{
  "name": "_buk.fdt",
  "length_in_bytes": 26891307173,
  "reused": false,
  "recovered_in_bytes": 35324928266,
  "recovered_from_snapshot_in_bytes": 8433621093
}
...
```

NB `35324928266 == 26891307173 + 8433621093`

This causes Elasticsearch to compute a negative value for the reserved space on the target node of this recovery, which prevents stats APIs from working. This appears to be caused by a network-level failure which triggers a retry without first cancelling an ongoing download of a different file with the same name from the snapshot repository.

---

### Workaround

Restart the target node to clear these stats.

### PR DESCRIPTION
**Pull request by fcofdez**
This commit fixes over-accounting recovered bytes when a file is recovered from a snapshot
and the connection between the source node and target node is closed. In that case the 
target keeps downloading the snapshot file and in some rare timings can lead to over-accounting
the number of recovered bytes. With this change when the recovery is restarted a new temporary
file prefix will be used, as well as a fresh `RecoveryState` avoiding collisions between the different
recovery phases.

Closes https://github.com/elastic/elasticsearch/issues/90441

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Relates (maybe closes?) https://github.com/elastic/elasticsearch/issues/90441

**Comment by fcofdez** [comment]
Yes, I wanted to run CI first to see how many things break before opening the final PR.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java [hunk 1]
```diff
@@ -128,6 +128,23 @@ public RecoveryTarget(
...
+        // cleaning itself once all the outstanding writes finish.
+        multiFileWriter.close();
+        final String tempFilePrefix = RECOVERY_PREFIX + UUIDs.randomBase64UUID() + ".";
+        this.multiFileWriter = new MultiFileWriter(
+            indexShard.store(),
+            indexShard.recoveryState().getIndex(),
+            tempFilePrefix,
+            logger,
+            this::ensureRefCount
+        );
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java:145]: Can we share this with the code in the constructor, looks identical - and is important that it is.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/recovery/SearchableSnapshotRecoveryState.java [hunk 2]
```diff
@@ -51,6 +51,12 @@ public synchronized RecoveryState setRemoteTranslogStage() {
         return super.setStage(Stage.FINALIZE);
     }
 
+    @Override
+    public synchronized RecoveryState reset() {
+        setStage(Stage.INIT);
+        return this;
```
Comment (henningandersen) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/recovery/SearchableSnapshotRecoveryState.java:57]: Would it not be nicer to also return a new instance here? Would that be problematic?

File: x-pack/plugin/snapshot-based-recoveries/src/internalClusterTest/java/org/elasticsearch/xpack/snapshotbasedrecoveries/recovery/SnapshotBasedIndexRecoveryIT.java [hunk 3]
```diff
@@ -1200,6 +1255,142 @@ public void testRecoveryRetryKeepsTheGrantedSnapshotFileDownloadPermit() throws
...
+        assertThat(connectionRefs, is(not(empty())));
+        connectionRefs.forEach(Transport.Connection::close);
+
+        CountDownLatch firstFileChunkSent = new CountDownLatch(1);
+        CountDownLatch blockFileChunkDownload = new CountDownLatch(1);
+
+        targetMockTransportService.addRequestHandlingBehavior(
+            PeerRecoveryTargetService.Actions.FILE_CHUNK,
+            (handler, request, channel, task) -> {
+                firstFileChunkSent.countDown();
```
Comment (henningandersen) [inline x-pack/plugin/snapshot-based-recoveries/src/internalClusterTest/java/org/elasticsearch/xpack/snapshotbasedrecoveries/recovery/SnapshotBasedIndexRecoveryIT.java:1366]: I am not sure I see how we are sure this has not completed already before adding the request handling behavior?

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 4]
```diff
@@ -2001,7 +2001,7 @@ public void resetRecoveryStage() {
         if (state != IndexShardState.RECOVERING) {
             throw new IndexShardNotRecoveringException(shardId, state);
         }
-        recoveryState().setStage(RecoveryState.Stage.INIT);
+        this.recoveryState = recoveryState().reset();
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:2004]: I wonder if we need to take the mutex around this? Did not find a specific issue though but seems we protect changes to the variable by `mutex`?

**Review by henningandersen — COMMENTED** [review]
This looks like a good direction, left a few smaller comments.

File: x-pack/plugin/snapshot-based-recoveries/src/internalClusterTest/java/org/elasticsearch/xpack/snapshotbasedrecoveries/recovery/SnapshotBasedIndexRecoveryIT.java (hunk 3 above)
Comment (fcofdez) [reply x-pack/plugin/snapshot-based-recoveries/src/internalClusterTest/java/org/elasticsearch/xpack/snapshotbasedrecoveries/recovery/SnapshotBasedIndexRecoveryIT.java:1366]: Yes, that's a bit racy. I'll move it before we close the connections. Good catch.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/recovery/SearchableSnapshotRecoveryState.java (hunk 2 above)
Comment (fcofdez) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/recovery/SearchableSnapshotRecoveryState.java:57]: For searchable snapshots we shouldn't run into this problem, we don't recover any file through regular recovery. So I figured that it should be fine to reuse the same instance.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 4 above)
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:2004]: I had the impression that we guard the `state` field with that mutex. But just to be safe I'll add the mutex guard here too.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java [hunk 5]
```diff
@@ -114,20 +114,28 @@ public RecoveryTarget(
...
-            this::ensureRefCount
-        );
+        this.multiFileWriter = createMultiFileWriter();
         this.store = indexShard.store();
         // make sure the store is not released until we are done.
         store.incRef();
         indexShard.recoveryStats().incCurrentAsTarget();
     }
 
+    private void refreshMultiFileWriter() {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java:124]: Can we call this "recreateMultiFileWriter" or "resetMultiFileWriter" instead? I think refresh does not match what it does.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java [hunk 6]
```diff
@@ -114,20 +114,28 @@ public RecoveryTarget(
...
         this.shardId = indexShard.shardId();
-        final String tempFilePrefix = RECOVERY_PREFIX + UUIDs.randomBase64UUID() + ".";
-        this.multiFileWriter = new MultiFileWriter(
-            indexShard.store(),
-            indexShard.recoveryState().getIndex(),
-            tempFilePrefix,
-            logger,
-            this::ensureRefCount
-        );
+        this.multiFileWriter = createMultiFileWriter();
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java:117]: Is this `MultiFileWriter` now ever used for anything but closing it? Perhaps we can check for null in `refreshMultiWriter` instead if I am right?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/recovery/SearchableSnapshotRecoveryState.java (hunk 2 above)
Comment (henningandersen) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/recovery/SearchableSnapshotRecoveryState.java:57]: To the reader however, it looks like this is either a micro-tuning (which seems unnecessary) or there is something special going on requiring us to not create a new state. I am fine to leave as is, but perhaps we can add a comment that it is equivalent to creating a new `SearchableSnapshotRecoveryState`?

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by fcofdez** [comment]
Thanks Henning!

**Pull request merged**